File: templates/python/spaghetti.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""个体轨迹 Spaghetti 图（纵向）+ 组均值 95%CI —— 卡片 04_时间趋势图/个体轨迹图_Spaghetti.md
合成纵向数据（每人多次随访），细线=个体，粗线=组均值。改 # >>> PARAM 区。"""
import os, sys, warnings; warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np, pandas as pd, matplotlib.pyplot as plt
from _common import set_theme, save_fig, pal
set_theme(); np.random.seed(7)

# >>> PARAM ------------------------------------------------------
NAME = "tpl_spaghetti"
TITLE = "Individual trajectories"; XLAB = "Visit"; YLAB = "Biomarker"
PALETTE = "cat_main"
FIGSIZE = (4.0, 3.0)
ID, T, Y, GROUP = "id", "visit", "y", "grp"
# 换真实数据： 长格式 df[ID,T,Y,GROUP]（每人多行）
rows = []
for g, (slope, base) in {"Treatment": (-0.6, 10), "Control": (-0.1, 10)}.items():
    for i in range(25):
        r0 = base + np.random.normal(0, 1.2); s = slope + np.random.normal(0, 0.25)
        for t in range(5):
            rows.append((f"{g}_{i}", t, r0 + s * t + np.random.normal(0, 0.4), g))
df = pd.DataFrame(rows, columns=[ID, T, Y, GROUP])
# <<< PARAM ------------------------------------------------------

colors = pal(PALETTE); levels = list(dict.fromkeys(df[GROUP]))
cmap = {lv: colors[i] for i, lv in enumerate(levels)}
fig, ax = plt.subplots(figsize=FIGSIZE)
for _, sub in df.groupby(ID):
    ax.plot(sub[T], sub[Y], color=cmap[sub[GROUP].iloc[0]], lw=.5, alpha=.22)
for g, sub in df.groupby(GROUP):
    m = sub.groupby(T)[Y].agg(["mean", "count", "std"])
    se = m["std"] / np.sqrt(m["count"])
    ax.fill_between(m.index, m["mean"] - 1.96 * se, m["mean"] + 1.96 * se, color=cmap[g], alpha=.2)
    ax.plot(m.index, m["mean"], color=cmap[g], lw=2.2, label=g)
ax.set_xlabel(XLAB); ax.set_ylabel(YLAB); ax.set_title(TITLE); ax.grid(False)
ax.legend(fontsize=8)
save_fig(fig, NAME)

File: templates/r/spaghetti.R
## 个体轨迹 Spaghetti 图（纵向）+ 组均值 95%CI —— 卡片 04_时间趋势图/个体轨迹图_Spaghetti.md
## 合成纵向数据，细线=个体，粗线=组均值。 Rscript templates/r/spaghetti.R
.args <- commandArgs(trailingOnly = FALSE)
.sd <- dirname(normalizePath(sub("^--file=", "", .args[grep("^--file=", .args)])))
source(file.path(.sd, "_common.R"))
suppressPackageStartupMessages(library(dplyr))
set.seed(2026)

# >>> PARAM ------------------------------------------------------
NAME <- "tpl_spaghetti"; TITLE <- "Individual trajectories"
XLAB <- "Visit"; YLAB <- "Biomarker"; PALETTE <- "cat_main"
# 换真实数据： 长格式 df(id, visit, y, grp)
mk <- function(g, slope, base, n = 25) do.call(rbind, lapply(1:n, function(i) {
  r0 <- base + rnorm(1, 0, 1.2); s <- slope + rnorm(1, 0, .25)
  data.frame(id = paste0(g, "_", i), visit = 0:4, y = r0 + s * (0:4) + rnorm(5, 0, .4), grp = g)
}))
df <- rbind(mk("Treatment", -0.6, 10), mk("Control", -0.1, 10))
# <<< PARAM ------------------------------------------------------

cols <- pal(PALETTE)[seq_along(unique(df$grp))]; names(cols) <- unique(df$grp)
agg <- df %>% group_by(grp, visit) %>%
  summarise(m = mean(y), se = sd(y) / sqrt(n()), .groups = "drop")
p <- ggplot() +
  geom_line(data = df, aes(visit, y, group = id, color = grp), linewidth = .25, alpha = .22) +
  geom_ribbon(data = agg, aes(visit, ymin = m - 1.96 * se, ymax = m + 1.96 * se, fill = grp), alpha = .2) +
  geom_line(data = agg, aes(visit, m, color = grp), linewidth = 1.1) +
  scale_color_manual(values = cols) + scale_fill_manual(values = cols) +
  labs(x = XLAB, y = YLAB, color = NULL, fill = NULL, title = TITLE)
save_gg(p, NAME, w = 100, h = 82)
